### PR TITLE
Cpu memory optimization

### PR DESCRIPTION
**Pull request by cehongwang**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/conversion/_conversion.py [hunk 1]
```diff
@@ -65,6 +70,29 @@ def interpret_module_to_result(
     )
 
     interpreter_result = interpreter.run()
+    # Delete the frozen parameters from the module to release CPU memory
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/_conversion.py:78]: Can we gate this by the same env variable as the malloc_trim?

File: py/torch_tensorrt/dynamo/conversion/_conversion.py [hunk 2]
```diff
@@ -104,7 +132,7 @@ def convert_module(
         )
 
     return rt_cls(
-        serialized_engine=interpreter_result.serialized_engine,
+        serialized_engine=interpreter_result.engine,
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/_conversion.py]: Will this be the serialized or live engine?

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 3]
```diff
@@ -733,6 +735,9 @@ def run(
                         return interpreter_result  # type: ignore[no-any-return]
 
         self._construct_trt_network_def()
+        _LOGGER.info(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py]: Debug level?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 4]
```diff
@@ -689,14 +690,17 @@ def compile(
 
     # Apply lowering on the graph module
     gm = post_lowering(gm, settings)
+    logger.debug(f"CPU memory usage after post_lowering: {get_cpu_memory_usage()} MB")
     logger.debug("Lowered Input graph: " + str(gm.graph))
 
     # Move the weights in the state_dict to CPU
     if offload_module_to_cpu:
+        deallocate_module(gm, delete_module=False)
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py:698]: Arent these the same?

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 5]
```diff
@@ -858,6 +862,11 @@ def preserve_module_specs(
     # Iterate over all components that can be accelerated
     # Generate the corresponding TRT Module for those
 
+    # Here we delete the frozen parameters from the graph module. Note this does not affect the submodules. We are going to delete the frozen parameters from the submodules in the convert_module function.
+    # This is done to release CPU memory.
+    for attr in dir(gm):
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/_compiler.py:867]: Lets make this opt in similar to malloc trim

**Review by narendasan — COMMENTED** [review]
We need the env variable control

File: py/torch_tensorrt/dynamo/_compiler.py (hunk 5 above)
Comment (cehongwang) [reply py/torch_tensorrt/dynamo/_compiler.py:867]: Shouldn't this be cleared no matter what?

File: py/torch_tensorrt/dynamo/conversion/_conversion.py (hunk 1 above)
Comment (cehongwang) [reply py/torch_tensorrt/dynamo/conversion/_conversion.py:78]: Do you mean something like
`export low_RAM_mode=1 python flux.py` or build env variable?

Comment (narendasan) [reply py/torch_tensorrt/dynamo/conversion/_conversion.py:78]: I would say something like `export TORCHTRT_USE_MALLOC_TRIM=1 python flux.py` and the goal would be for this to go away in 2.11 and be merged into `prioritize_host_memory_consumption=True`

File: py/torch_tensorrt/dynamo/partitioning/_adjacency_partitioner.py [hunk 6]
```diff
@@ -225,13 +226,80 @@ def partition_graph(self) -> torch.fx.GraphModule:
...
+        subgraphs = self.break_subgraphs(subgraphs, num_of_break=num_of_break)
+
         # Set the number of TRT engines to be generated
         self.num_trt_accelerated_subgraphs = len([s for s in subgraphs if s.is_acc])
 
         # Tag the accelerated nodes and split the graph accordingly
         self.tag(subgraphs)
         return self.split()
 
+    def calculate_num_of_break(self, subgraphs: List[Subgraph]) -> int:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/partitioning/_adjacency_partitioner.py:239]: I feel like this is too much of an heuristic based system. A better approach IMO is to calculate a graph size budget based on available memory (or eventually this could be user specified). Then of the TRT blocks we estimate its size and then decide how many subgraphs it should be split into to meet the budget

File: py/torch_tensorrt/dynamo/partitioning/_adjacency_partitioner.py [hunk 7]
```diff
@@ -225,13 +226,80 @@ def partition_graph(self) -> torch.fx.GraphModule:
...
+            [node for node in self.acc_nodes if "scaled_dot" in str(node.target)]
+        )
+        break_period = num_of_sdpa_node // num_of_break + 1
+        current_break_idx = 0
+        current_num_break = 0
+        new_subgraphs = []
+        for subgraph in subgraphs:
+            if subgraph.is_acc:
+                for i, node in enumerate(subgraph.nodes):
+                    if "scaled_dot" in str(node.target):
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/partitioning/_adjacency_partitioner.py:280]: Its fine if we do this for testing but really we should be taking a much more generic approach rather than assuming only sdpa is a viable break point.

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 8]
```diff
@@ -750,16 +755,19 @@ def run(
         self._create_timing_cache(
             builder_config, self.compilation_settings.timing_cache_path
         )
-        serialized_engine = self.builder.build_serialized_network(
+
+        cuda_engine = self.builder.build_engine_with_config(
```
Comment (lanluo-nvidia) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:759]: Curious what is the benefit in CPU memory, if we return cuda_engine instead of serialized_engine?
because interpret_module_to_result() will anyway deserialize the cuda_engine to serialized_engine str.
